draft-cpbs-pseudorandom-ctls.md

```diff
@@ -90,60 +98,81 @@ A cTLS template is structured as a JSON object.  This extension is represented b
 
 ~~~
-STPRP-Encipher(key, tweak, message) -> ciphertext
-STPRP-Decipher(key, tweak, ciphertext) -> message
+TSPRP-Encipher(key, tweak, message) -> ciphertext
+TSPRP-Decipher(key, tweak, ciphertext) -> message
+Twist(input) -> twisted
+Untwist(twisted) -> input
 ~~~
 
-The STPRP specifies the length (in bytes) of the key.  The tweak is a byte string of any length.  The STPRP uses the key and tweak to encipher the input message, which also may have any length.  The output ciphertext has the same length as the input message.
+> QUESTION: Should we allow a key or tweak for the twist?
 
-The Pseudorandom cTLS design assumes that the negotiated AEAD algorithm produces pseudorandom ciphertexts.  This is not a requirement of the AEAD specification {{!RFC5116}}, but it is true of popular AEAD algorithms like AES-GCM and ChaCha20-Poly1305.
+The cipher suite specifies the length (in bytes) of its key.  The tweak is a byte string of any length.  Both the TSPRP and "twist" accept any input with length >= 16 bytes.
 
+In order to satisfy the requirements ({{requirements}}), `TSPRP-Encipher(key, tweak, Twist(AEAD-Encrypt(key2, nonce, plaintext)))` must be an Everywhere-Preimage-Resistant hash {{?EPRE=DOI.10.1007/978-3-540-25937-4_24}} of `plaintext` for any AEAD family.
 
-Pseudorandom cTLS uses the STPRP to encipher all plaintext handshake records, including the record headers.  As long as there is sufficient entropy in the `key_share` extension or `random` field of the ClientHello (resp. ServerHello) the STPRP output will be pseudorandom.
+> TODO: Formalize this notion better.
 
-> TODO: Check that the assumptions hold for HelloRetryRequest.  As long as no handshake messages are repeated verbatim, it should be fine, but we need to check whether an active attacker can trigger a replay.
+### The "hctr2-sha256" cipher suite
 
-Pseudorandom cTLS also enciphers every record header.  In addition to the header, 16 bytes of the AEAD ciphertext itself is enciphered to ensure the input has enough entropy.  All currently registered AEAD algorithms produce at least this much ciphertext from any input.  Any AEAD algorithm that can produce smaller ciphertexts is not compatible with this specification.
+HCTR2 is a fast TSPRP based on AES and a polynomial MAC (closely related to the GMAC used in AES-GCM) (https://eprint.iacr.org/2021/1441).  However, HCTR2 alone is not sufficient for Pseudorandom cTLS.  For example, with knowledge of the keys, it is likely possible to control much of the output of `HCTR2(AES-GCM(plaintext))`.
 
-### With Streaming Transports
+To provide a suitable cipher suite, we define the "twist" function as
 
-When used over a streaming transport, Pseudorandom cTLS requires that headers have predictable lengths.  This creates the following limitations:
+~~~python
+hash = SHA256(input[:-16])
+twisted = input[:-16] + XOR(input[-16:], hash[:16])
```

Comment:
Another possibility that I think would work would be
~~~python
if len(input) <= 32: return input
hash = SHA256(input[:-32])
twisted = input[:-32] + XOR(input[-32:], hash)
~~~

That introduces two cases, which complicates the analysis, but it might be still be easier to analyze than truncating the hash.